### PR TITLE
Remove propTypes from production builds

### PR DESCRIPTION
**Pull request by mbrevda**
Motivation
===
Removed propTypes from production builds, as they are unused by React in production and just take up extra bytes. This can reduce the minimized dist build by as much as **~3%!**:

```
// before
48170 Oct 19 11:21 react-intl.min.js

// after
46836 Oct 19 11:24 react-intl.min.js
```

What changed
===
For `lib/` builds, propTypes are wrapped in a conditional statement (`if(process.env.NODE_ENV !== 'production'){}`). This keeps the proptypes in place during development so that dev have access to them, but removes them when built with a packager (browserify/webpack/rollup/etc) that replaces `process.env.NODE_ENV` with the actual value.

For `dist/` build, the process is similar, but when `isProduction == true`, the propTypes will be completely removed. It was necessary to move the `replace()` plugin to the `isProduction == true` build so that propTypes aren't completely removed for non production build.

Finally, due to the non-standard way of how propTypes are defined in this project, the definitions themselves were manually wrapped where possible (see below). This, again, keeps them in development build but removes them for production builds.

TODO:
===
It seems some proptypes are used by the code during runtime for non-propType related activity. Can these be updated/removed/otherwise minified for even more saving?

Disclaimer
===
I hope I didn't misunderstand how the build process works!

Prior art:
===
 * react-virtualized: https://github.com/bvaughn/react-virtualized/pull/438 & https://github.com/bvaughn/react-virtualized/pull/440
 * react-router: https://github.com/ReactTraining/react-router/issues/4046
 * react: https://github.com/facebook/react/pull/7651, https://github.com/facebook/react/pull/2948, amongst others
 * the plugin: https://github.com/oliviertassinari/babel-plugin-transform-react-remove-prop-types

Related:
===
This is a first step towards https://github.com/yahoo/react-intl/issues/687

### CODE REVIEW

**Comment by yahoocla** [comment]
Thank you for submitting this pull request, however I do not see a valid CLA on file for you. Before we can merge this request please visit https://yahoocla.herokuapp.com/ and agree to the terms. Thanks! :smile:

**Comment by yahoocla** [comment]
CLA is valid!

File: scripts/build-dist.js [hunk 1]
```diff
@@ -57,13 +59,13 @@ let plugins = [
...
     }),
-    replace({
-        'process.env.NODE_ENV': JSON.stringify(process.env.NODE_ENV),
-    }),
 ];
 
 if (isProduction) {
     plugins.push(
+        replace({
+            'process.env.NODE_ENV': JSON.stringify(process.env.NODE_ENV),
```
Comment (ericf) [inline scripts/build-dist.js]: What did you move this into the the production build only?

File: scripts/build-dist.js [hunk 2]
```diff
@@ -45,6 +45,8 @@ babelConfig.babelrc = false;
 babelConfig.presets = babelConfig.presets.map((preset) => {
     return preset === 'es2015' ? 'es2015-rollup' : preset;
 });
+babelConfig.plugins = babelConfig.plugins || [];
+babelConfig.plugins.push(['transform-react-remove-prop-types', {'mode': 'wrap'}]);
```
Comment (ericf) [inline scripts/build-dist.js]: I think a better approach would be to add this to the `.babelrc` when `"env"` is `"production"`.

Comment (ericf) [inline scripts/build-dist.js]: Why `mode: 'wrap'`, instead of the default to remove?

**Comment by mbrevda** [comment]
> Yes, I'd like the leave the types.js file alone for this reason.

Leaving types.js alone mostly inverses the (already tiny) gains we are seeking here. I suggest modifying it smartly and that which can be eliminates should be, as I have done in the PR.

> I'd like to know what the gzip savings is of this change

I don't know. Even gzipped code has a price during unzipping and parsing. While these changes probably won't add much weight in either direction, every byte counts.

> Does this Babel plugin support removing static proptType = {}; from the components? 

yes, it does. However some have argued that leaving them in during development is valuable to developers (otherwise, why include them at all?). The middle of the road seems to be: ship them so that dev's can leverage them, but add a simple way to strip them out for production (see links above for more).

> allow Rollup to do its tree-shaking magic

As I mentioned in #687, tree-shaking is more smoke and mirrors then magic. While it mostly delivers for library authors, it really disappoints when working on a real life project. Currently the react project advocates the `process.env.NODE_ENV !== 'production'` method (and makes heavy use of it, making that the lowest common denominator for ensuring the minimal amount of code is being shipped.

File: scripts/build-dist.js (hunk 1 above)
Comment (mbrevda) [inline scripts/build-dist.js]: see https://github.com/yahoo/react-intl/pull/696#issuecomment-254942556, 3rd quote

File: scripts/build-dist.js (hunk 2 above)
Comment (mbrevda) [inline scripts/build-dist.js]: We could move it to `.babelrc`, but it doesn't seem like your currently setup for env builds -  and thats beyond the scope of this PR.

Comment (ericf) [inline scripts/build-dist.js]: The scripts are doing builds for both development and production. I'm suggesting that you do what's in the README for this Babel plugin: https://github.com/oliviertassinari/babel-plugin-transform-react-remove-prop-types#via-babelrc-recommended

**Comment by ericf** [comment]
I quickly tried what I suggested, which required no code changes, just `package.json` and `.babelrc` changes and got these results:

**Before:**
```
48170 Oct 19 18:45 react-intl.min.js
13363 Oct 19 18:45 react-intl.min.js.gz
```

**After:**
```
47025 Oct 19 18:41 react-intl.min.js
13206 Oct 19 18:41 react-intl.min.js.gz
```

So **2.4%** savings on `.min.js` and **1.2%** savings on `.min.js.gz`.

As for reducing `src/types.js` in production, I have other requests asking to _expose_ that entire module for them to use: https://github.com/yahoo/react-intl/issues/549

Here's my diff:
```diff
diff --git a/package.json b/package.json
index 7b044b4..8721ab7 100644
--- a/package.json
+++ b/package.json
@@ -65,6 +65,7 @@
     "babel-plugin-transform-es3-member-expression-literals": "^6.3.13",
     "babel-plugin-transform-es3-property-literals": "^6.3.13",
     "babel-plugin-transform-object-rest-spread": "^6.1.18",
+    "babel-plugin-transform-react-remove-prop-types": "^0.2.10",
     "babel-polyfill": "^6.3.14",
     "babel-preset-es2015": "^6.1.18",
     "babel-preset-es2015-rollup": "^1.1.1",
diff --git a/src/.babelrc b/src/.babelrc
index b80ddf1..8412546 100644
--- a/src/.babelrc
+++ b/src/.babelrc
@@ -14,6 +14,11 @@
             "plugins": [
                 "istanbul"
             ]
+        },
+        "production": {
+            "plugins": [
+                "transform-react-remove-prop-types"
+            ]
         }
     }
 }
```

**Comment by ericf** [comment]
btw, Rollup is smart enough to tree-shake the defs in `src/types.js` that aren't being used in the prod code; e.g. `messageDescriptorPropTypes`.

**Comment by ericf** [comment]
You could also define a few of the repetitive types so they don't need to be duplicated in `src/types.js`:
```diff
diff --git a/src/types.js b/src/types.js
index ffad182..12d3a1f 100644
--- a/src/types.js
+++ b/src/types.js
@@ -40,26 +40,30 @@ export const messageDescriptorPropTypes = {
     defaultMessage: string,
 };
 
+const localeMatcher = oneOf(['best fit', 'lookup']);
+const narrowShortLong = oneOf(['narrow', 'short', 'long']);
+const numeric2digit = oneOf(['numeric', '2-digit']);
+
 export const dateTimeFormatPropTypes = {
-    localeMatcher: oneOf(['best fit', 'lookup']),
+    localeMatcher: localeMatcher,
     formatMatcher: oneOf(['basic', 'best fit']),
 
     timeZone: string,
     hour12  : bool,
 
-    weekday     : oneOf(['narrow', 'short', 'long']),
-    era         : oneOf(['narrow', 'short', 'long']),
-    year        : oneOf(['numeric', '2-digit']),
+    weekday     : narrowShortLong,
+    era         : narrowShortLong,
+    year        : numeric2digit,
     month       : oneOf(['numeric', '2-digit', 'narrow', 'short', 'long']),
-    day         : oneOf(['numeric', '2-digit']),
-    hour        : oneOf(['numeric', '2-digit']),
-    minute      : oneOf(['numeric', '2-digit']),
-    second      : oneOf(['numeric', '2-digit']),
+    day         : numeric2digit,
+    hour        : numeric2digit,
+    minute      : numeric2digit,
+    second      : numeric2digit,
     timeZoneName: oneOf(['short', 'long']),
 };
 
 export const numberFormatPropTypes = {
-    localeMatcher: oneOf(['best fit', 'lookup']),
+    localeMatcher: localeMatcher,
 
     style          : oneOf(['decimal', 'currency', 'percent']),
     currency       : string,
```

**Comment by mbrevda** [comment]
I have made some of the suggested changes. Were currently looking at:
```
46732 Oct 20 10:40 react-intl.min.js
13123 Oct 20 10:40 react-intl.min.js.gz
```

What other changes do you feel are necessary or can be usefull?

As file compression works better on larger datasets, the gz size isn't fully indicative (as I'm sure you know).

File: src/types.js [hunk 3]
```diff
@@ -7,59 +7,62 @@
...
+const localeMatcher = oneOf(['best fit', 'lookup']);
+const narrowShortLong = oneOf(['narrow', 'short', 'long']);
+const numeric2digit = oneOf(['numeric', '2-digit']);
+const funcReq = func.isRequired;
 
 export const intlConfigPropTypes = {
     locale  : string,
     formats : object,
     messages: object,
-
```
Comment (ericf) [inline src/types.js]: Please add back the line break here.

File: src/types.js [hunk 4]
```diff
@@ -7,59 +7,62 @@
...
 
-export const intlFormatPropTypes = {
-    formatDate       : func.isRequired,
-    formatTime       : func.isRequired,
-    formatRelative   : func.isRequired,
-    formatNumber     : func.isRequired,
-    formatPlural     : func.isRequired,
-    formatMessage    : func.isRequired,
-    formatHTMLMessage: func.isRequired,
+export const intlFormatPropTypes =  {
```
Comment (ericf) [inline src/types.js]: Remove the added space.

File: src/types.js [hunk 5]
```diff
@@ -7,59 +7,62 @@
...
+    formatTime       : funcReq,
+    formatRelative   : funcReq,
+    formatNumber     : funcReq,
+    formatPlural     : funcReq,
+    formatMessage    : funcReq,
+    formatHTMLMessage: funcReq,
 };
 
-export const intlShape = shape({
+export const intlShape = process.env.NODE_ENV !== 'production' ? shape({
```
Comment (ericf) [inline src/types.js]: Please remove these ternaries for the reasons stated in the other comments.

File: src/types.js [hunk 6]
```diff
@@ -7,59 +7,62 @@
...
     ...intlConfigPropTypes,
     ...intlFormatPropTypes,
     formatters: object,
-    now: func.isRequired,
-});
+    now: funcReq,
+}) : ()=>{};
 
-export const messageDescriptorPropTypes = {
+export const messageDescriptorPropTypes = process.env.NODE_ENV !== 'production' ? {
```
Comment (ericf) [inline src/types.js]: Please remove these ternaries for the reasons stated in the other comments.

File: src/types.js [hunk 7]
```diff
@@ -73,11 +76,11 @@ export const numberFormatPropTypes = {
     maximumSignificantDigits: number,
 };
 
-export const relativeFormatPropTypes = {
+export const relativeFormatPropTypes = process.env.NODE_ENV !== 'production' ? {
```
Comment (ericf) [inline src/types.js]: Please remove these ternaries for the reasons stated in the other comments.

File: src/types.js [hunk 8]
```diff
@@ -73,11 +76,11 @@ export const numberFormatPropTypes = {
...
     style: oneOf(['best fit', 'numeric']),
     units: oneOf(['second', 'minute', 'hour', 'day', 'month', 'year']),
-};
+} : {};
 
-export const pluralFormatPropTypes = {
+export const pluralFormatPropTypes = process.env.NODE_ENV !== 'production' ? {
```
Comment (ericf) [inline src/types.js]: Please remove these ternaries for the reasons stated in the other comments.

**Comment by mbrevda** [comment]
> I was literally stating a fact

Got it!

Did I miss anything else?

File: scripts/build-dist.js (hunk 1 above)
Comment (ericf) [inline scripts/build-dist.js]: @mbrevda can you revert this change to `scripts/built-dist.js` too?

**Comment by ericf** [comment]
> Did I miss anything else?

@mbrevda can you revert the changes to `scripts/build-dist.js` as well? https://github.com/yahoo/react-intl/pull/696#discussion_r84331248

**Comment by mbrevda** [comment]
I'm pretty sure that would also prevent propTypes from being utilized by
devs during development. Are you certain you want me to revert the change?

**Comment by ericf** [comment]
> I'm pretty sure that would also prevent propTypes from being utilized by
> devs during development. Are you certain you want me to revert the change?

They are there in development. And you already reverted the change to `scripts/build-dist.js` before I merged this.

**Comment by mbrevda** [comment]
I'm pretty sure that would also prevent propTypes from being utilized by
devs during development - the only time propTypes could be used. Are you
certain you want that reverted?

On Thu, Oct 20, 2016, 9:30 PM Eric Ferraiuolo <notifications@github.com>
wrote:

Did I miss anything else?

@mbrevda <https://github.com/mbrevda> can you revert the changes to
scripts/build-dist.js as well? #696 (comment)
<https://github.com/yahoo/react-intl/pull/696#discussion_r84331248>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/yahoo/react-intl/pull/696#issuecomment-255189182>, or mute
the thread

**Comment by mbrevda** [comment]
Hmm - seems github just posted emails sent a week ago. Ignore them!

**Comment by ericf** [comment]
Ha, lol, I was wondering what was going on 😄
